src/interface/sparse_pack_base.hpp

```diff
@@ -0,0 +1,199 @@
+  std::vector<MetadataFlag> flags;
+  bool with_fluxes;
+  bool coarse;
+
+  PackDescriptor(const std::vector<std::string> &vars, const std::vector<bool> &use_regex,
+                 const std::vector<MetadataFlag> &flags, bool with_fluxes, bool coarse)
+      : vars(vars), use_regex(use_regex), flags(flags), with_fluxes(with_fluxes),
+        coarse(coarse) {
+    PARTHENON_REQUIRE(use_regex.size() == vars.size(),
+                      "Must have a regex flag for each variable.");
+    PARTHENON_REQUIRE(!(with_fluxes && coarse),
+                      "Probably shouldn't be making a coarse pack with fine fluxes.");
+  }
+
+  PackDescriptor(const std::vector<std::pair<std::string, bool>> &vars_in,
+                 const std::vector<MetadataFlag> &flags, bool with_fluxes, bool coarse)
+      : flags(flags), with_fluxes(with_fluxes), coarse(coarse) {
+    for (auto var : vars_in) {
+      vars.push_back(var.first);
+      use_regex.push_back(var.second);
+    }
+    PARTHENON_REQUIRE(!(with_fluxes && coarse),
+                      "Probably shouldn't be making a coarse pack with fine fluxes.");
+  }
+
+  PackDescriptor(const std::vector<std::string> &vars_in,
+                 const std::vector<MetadataFlag> &flags, bool with_fluxes, bool coarse)
+      : vars(vars_in), use_regex(vars_in.size(), false), flags(flags),
+        with_fluxes(with_fluxes), coarse(coarse) {
+    PARTHENON_REQUIRE(!(with_fluxes && coarse),
+                      "Probably shouldn't be making a coarse pack with fine fluxes.");
+  }
+};
+} // namespace impl
+
+using namespace impl;
+
+class PackIdx {
+  std::size_t vidx;
+  int off;
+
+ public:
+  KOKKOS_INLINE_FUNCTION
+  explicit PackIdx(std::size_t var_idx) : vidx(var_idx), off(0) {}
+  KOKKOS_INLINE_FUNCTION
+  PackIdx &operator=(std::size_t var_idx) {
+    vidx = var_idx;
+    off = 0;
+    return *this;
+  }
+  KOKKOS_INLINE_FUNCTION
+  PackIdx(std::size_t var_idx, int off) : vidx(var_idx), off(off) {}
+  KOKKOS_INLINE_FUNCTION
+  std::size_t Vidx() { return vidx; }
+  KOKKOS_INLINE_FUNCTION
+  int Off() { return off; }
+};
+
+template <class T, class = typename std::enable_if<std::is_integral<T>::value>::type>
+KOKKOS_INLINE_FUNCTION PackIdx operator+(PackIdx idx, T off) {
+  return PackIdx(idx.Vidx(), idx.Off() + off);
+}
+
+template <class T, class = typename std::enable_if<std::is_integral<T>::value>::type>
+KOKKOS_INLINE_FUNCTION PackIdx operator+(T off, PackIdx idx) {
+  return idx + off;
+}
+
+using SparsePackIdxMap = std::unordered_map<std::string, std::size_t>;
+
+class SparsePackBase {
+ protected:
+  friend class SparsePackCache;
+
+  using test_func_t =
+      std::function<bool(int, const std::shared_ptr<CellVariable<Real>> &)>;
+  using pack_t = ParArray3D<ParArray3D<Real>>;
+  using bounds_t = ParArray3D<int>;
+  using alloc_t = std::vector<bool>;
+  using coords_t = ParArray1D<ParArray0D<Coordinates_t>>;
+
+  pack_t pack_;
+  bounds_t bounds_;
+  coords_t coords_;
+
+  bool with_fluxes_;
```

Comment:
I don't know how I feel about the fluxpack being the same type as the variable pack. I *think* I like it.

Reply:
Unless there is a performance hit I think I like it, it certainly reduces code complexity in my mind.